ext/semian_cb_data/semian_cb_data.c

```diff
@@ -0,0 +1,900 @@
+{
+  /* noop */
+}
+
+static void
+semian_cb_data_free(void *ptr)
+{
+  semian_cb_data *data = (semian_cb_data *) ptr;
+
+
+  // Under normal circumstances, memory use should be in the order of bytes,
+  //   and shouldn't increase if the same key/id is used
+  //   so there is no need to call this unless certain all other semian processes are stopped
+  //   (also raises concurrency errors: "object allocation during garbage collection phase")
+
+  //semian_cb_data_delete_memory_inner (data);
+
+  xfree(data);
+}
+
+static size_t
+semian_cb_data_memsize(const void *ptr)
+{
+  return sizeof(semian_cb_data);
+}
+
+static VALUE
+semian_cb_data_alloc(VALUE klass)
+{
+  VALUE obj;
+  semian_cb_data *ptr;
+
+  obj = TypedData_Make_Struct(klass, semian_cb_data, &semian_cb_data_type, ptr);
+  return obj;
+}
+
+
+
+
+
+
+/*
+ * Init function exposed as ._initialize() that is delegated by .initialize()
+ */
+static VALUE
+semian_cb_data_init(VALUE self, VALUE id, VALUE size, VALUE permissions)
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct(self, semian_cb_data, &semian_cb_data_type, ptr);
+
+  if (TYPE(id) != T_SYMBOL && TYPE(id) != T_STRING)
+    rb_raise(rb_eTypeError, "id must be a symbol or string");
+  if (TYPE(size) != T_FIXNUM /*|| TYPE(size) != T_BIGNUM*/)
+    rb_raise(rb_eTypeError, "expected integer for arr_max_size");
+  if (TYPE(permissions) != T_FIXNUM)
+    rb_raise(rb_eTypeError, "expected integer for permissions");
+
+  if (NUM2SIZET(size) <= 0)
+    rb_raise(rb_eArgError, "arr_max_size must be larger than 0");
+
+  const char *id_str = NULL;
+  if (TYPE(id) == T_SYMBOL) {
+    id_str = rb_id2name(rb_to_id(id));
+  } else if (TYPE(id) == T_STRING) {
+    id_str = RSTRING_PTR(id);
+  }
+  ptr->key = generate_key(id_str);
+  //rb_warn("converted name %s to key %d", id_str, ptr->key);
+
+  // Guarantee arr_max_size >=1 or error thrown
+  ptr->arr_max_size = NUM2SIZET(size);
+
+  // id's default to -1
+  ptr->semid = -1;
+  ptr->shmid = -1;
+  // addresses default to NULL
+  ptr->shm_address = 0;
+  ptr->lock_triggered = false;
+
+  semian_cb_data_acquire_semaphore(self, permissions);
+  semian_cb_data_acquire_memory(self, permissions);
+
+  return self;
+}
+
+static VALUE
+semian_cb_data_clean(VALUE self)
+{
+  semian_cb_data_delete_memory(self);
+  semian_cb_data_delete_semaphore(self);
+  return self;
+}
+
+
+
+/*
+ * Functions set_semaphore_permissions, configure_tickets, create_semaphore
+ * are taken from semian.c with extra code removed
+ */
+static void
+set_semaphore_permissions(int sem_id, int permissions)
+{
+  union semun sem_opts;
+  struct semid_ds stat_buf;
+
+  sem_opts.buf = &stat_buf;
+  semctl(sem_id, 0, IPC_STAT, sem_opts);
+  if ((stat_buf.sem_perm.mode & 0xfff) != permissions) {
+    stat_buf.sem_perm.mode &= ~0xfff;
+    stat_buf.sem_perm.mode |= permissions;
+    semctl(sem_id, 0, IPC_SET, sem_opts);
+  }
+}
+
+
+static void
+configure_tickets(int sem_id, int tickets, int should_initialize)
```

Comment:
Ditto with borrowing from `semian.c` with a flag to trigger the fancy 3-semaphore delta dance.